File: python/neural_network.py
import numpy as np
import math 
class NeuralNetwork:
    inputN  = None
    hiddenN  = None
    outputN  = None
    learningRate  = None
    tolerance = None
    _input  = None
    output  = None
    target  = None
    hidden  = None
    epoc = None
    weights = []
    biases = []
    
    def __init__(self, inputN, hiddenN, outputN, learningRate, tolerance, epoc):
        self.inputN = inputN
        self.hiddenN = hiddenN
        self.outputN = outputN
        self.learningRate = learningRate
        self.tolerance = tolerance
        self.epoc = epoc

        self._input = np.random.uniform(0,1,[inputN,1])
        self.hidden= np.random.uniform(0,1,[hiddenN,1])
        self.output = np.random.uniform(0,1,[outputN,1])

        self.weights.append(np.random.uniform(0,1,[hiddenN,inputN]))
        self.weights.append(np.random.uniform(0,1,[outputN,hiddenN]))
        self.biases.append(np.random.uniform(0,1,[hiddenN,1]))
        self.biases.append(np.random.uniform(0,1,[outputN,1]))
    def activation(self, temp):
        for i in range(len(temp)):
            for j in range(len(temp[0])):
                 temp[i][j]= (1/(1+math.exp(-1*temp[i][j])))
    
    def error(self, output, target):
        err = 0.0
        temp = np.empty([len(output),len(output[0])])
        for i in range(len(temp)):
            for j in range(len(temp[0])):
                temp[i][j]=0.5*(math.pow((target[i][j]-output[i][j]),2))
        for i in range(len(temp)):
            err+=temp[i][0]
        return err
    
    def dError(self, output, target):
        temp = np.subtract(output, target)
        return temp

    def dActivation(self, output):
        temp = np.empty([len(output),len(output[0])])
        for i in range(len(temp)):
            for j in range(len(temp[0])):
                temp[i][j] = output[i][j] * (1 - output[i][j])
        return temp

    def feedForward(self):
        self.hidden = self.weights[0].dot(self._input)#h
        self.hidden = np.add(self.hidden, self.biases[0])#neth
        self.activation(self.hidden)#outh
        self.output = self.weights[1].dot(self.hidden)#o
        self.output = np.add(self.output,self.biases[1])#neto     
        self.activation(self.output)#outo 

    def backProp(self):
        #output layer
        out_d_error_L2 = self.dError(self.output, self.target)
        net_d_out_L2 = self.dActivation(self.output)
        net_d_error_L2 = np.multiply(out_d_error_L2, net_d_out_L2)
        w_d_error_L2 = net_d_error_L2.dot(np.transpose(self.hidden))
        out_d_error_L1 = np.transpose(self.weights[1]).dot(net_d_error_L2)
        self.weights[1]=np.subtract(self.weights[1], np.multiply(w_d_error_L2, self.learningRate))
        
        #hidden layer
        net_d_out_L1 = self.dActivation(self.hidden)
        net_d_error_L1 = np.multiply(out_d_error_L1, net_d_out_L1)
        w_d_error_L1 = net_d_error_L1.dot(np.transpose(self._input))
        self.weights[0]=np.subtract(self.weights[0], np.multiply(w_d_error_L1, self.learningRate))


nn = NeuralNetwork(2,2,1,0.5,0.05,1200)
inputs = []
targets = []
outputs = []

inputs.append(np.empty([nn.inputN,1]))
inputs.append(np.empty([nn.inputN,1]))
inputs.append(np.empty([nn.inputN,1]))
inputs.append(np.empty([nn.inputN,1]))

targets.append(np.empty([nn.outputN,1]))
targets.append(np.empty([nn.outputN,1]))
targets.append(np.empty([nn.outputN,1]))
targets.append(np.empty([nn.outputN,1]))

outputs.append(np.empty([nn.outputN,1]))
outputs.append(np.empty([nn.outputN,1]))
outputs.append(np.empty([nn.outputN,1]))
outputs.append(np.empty([nn.outputN,1]))

inputs[0][0][0]=0
inputs[0][1][0]=0
inputs[1][0][0]=0
inputs[1][1][0]=1
inputs[2][1][0]=1
inputs[2][0][0]=0
inputs[3][0][0]=1
inputs[3][1][0]=1

targets[0][0][0]=0
targets[1][0][0]=1
targets[2][0][0]=1
targets[3][0][0]=0

i=0
while True:
    for j in range(len(inputs)):
        
        nn._input=inputs[j]
        nn.target=targets[j]
        nn.feedForward()
        print(str(j) + " "  + str(nn.output))
        outputs[j]=nn.output
        nn.backProp()
    print("epoc")
    print(i)
    if abs(outputs[0]-targets[0])<nn.tolerance and abs(outputs[1]-targets[1])<nn.tolerance and abs(outputs[2]-targets[2])<nn.tolerance and abs(outputs[3]-targets[3])<nn.tolerance:
        break
    i+=1

File: c#/NeuralNetwork.cs
using System;

namespace neuralnetwork {



    /**
     *
     * @author Ilterkaan Karaca
     */
    public class NeuralNetwork
    {
        int inputN, hiddenN, outputN;
        float learningRate;
        Matrix input;
        Matrix output;
        Matrix target;
        Matrix hidden;
        Matrix[] weights = new Matrix[2];
        Matrix[] biases = new Matrix[2];
        public NeuralNetwork(int inputN, int hiddenN, int outputN, float learningRate)
        {
            this.inputN = inputN;
            this.hiddenN = hiddenN;
            this.outputN = outputN;
            this.learningRate = learningRate;

            input = new Matrix(inputN, 1);
            hidden = new Matrix(hiddenN, 1);
            output = new Matrix(outputN, 1);

            weights[0] = new Matrix(hiddenN, inputN);
            weights[1] = new Matrix(outputN, hiddenN);
            biases[0] = new Matrix(hiddenN, 1);
            biases[1] = new Matrix(outputN, 1);


            weights[0].randomize();
            weights[1].randomize();
            biases[0].randomize();
            biases[1].randomize();

        }
        public static void activation(Matrix temp)
        {
            for (int i = 0; i < temp.matrix.length; i++)
            {
                for (int j = 0; j < temp.matrix[0].length; j++)
                    temp.matrix[i][j] = (1f / (1f + (float)Math.exp(-temp.matrix[i][j])));
            }
        }
        public static float error(Matrix output, Matrix target)
        {
            float err = 0.0f;
            Matrix temp = new Matrix(output.row, output.column);
            for (int i = 0; i < temp.matrix.length; i++)
            {
                for (int j = 0; j < temp.matrix[0].length; j++)
                    temp.matrix[i][j] = 0.5f * (float)(Math.pow((double)(target.matrix[i][j] - output.matrix[i][j]), 2.0));
            }
            for (int i = 0; i < temp.matrix.length; i++)
            {
                err += temp.matrix[i][0];
            }
            return err;
        }

        public static Matrix dError(Matrix output, Matrix target)
        {
            Matrix temp = Matrix.sub(output, target);
            return temp;
        }
        public static Matrix dActivation(Matrix output)
        {
            Matrix temp = new Matrix(output.row, output.column);
            for (int i = 0; i < temp.matrix.length; i++)
            {
                for (int j = 0; j < temp.matrix[0].length; j++)
                    temp.matrix[i][j] = output.matrix[i][j] * (1f - output.matrix[i][j]);
            }
            return temp;
        }
        public void feedForward()
        {


            hidden = Matrix.mult(weights[0], input);//h

            hidden = Matrix.add(hidden, biases[0]);//neth

            activation(hidden);//outh

            output = Matrix.mult(weights[1], hidden);//o

            output = Matrix.add(output, biases[1]);//neto
            activation(output);//outo  

        }
        public void backProp()
        {
            //output layer
            Matrix out_d_error_L2 = dError(output, target);
            Matrix net_d_out_L2 = dActivation(output);
            Matrix net_d_error_L2 = Matrix.eWMult(out_d_error_L2, net_d_out_L2);
            Matrix w_d_error_L2 = Matrix.mult(net_d_error_L2, Matrix.transpose(hidden));
            Matrix out_d_error_L1 = Matrix.mult(Matrix.transpose(weights[1]), net_d_error_L2);
            weights[1] = Matrix.sub(weights[1], Matrix.scalarMult(w_d_error_L2, learningRate));

            //hidden layer
            Matrix net_d_out_L1 = dActivation(hidden);
            Matrix net_d_error_L1 = Matrix.eWMult(out_d_error_L1, net_d_out_L1);
            Matrix w_d_error_L1 = Matrix.mult(net_d_error_L1, Matrix.transpose(input));
            weights[0] = Matrix.sub(weights[0], Matrix.scalarMult(w_d_error_L1, learningRate));
        }
        public void Test()
        {
            Matrix temp = Matrix.transpose(this.weights[1]);
            Console.WriteLine(this.weights[1].matrix[1].length);
        }
        /**
         * @param args the command line arguments
         */
        public static void main(String[] args)
        {
            int epoc = 1;

            NeuralNetwork nn = new NeuralNetwork(2, 2, 1, 0.5f);
            Matrix[] inputs = new Matrix[4];
            Matrix[] targets = new Matrix[4];

            inputs[0] = new Matrix(nn.inputN, 1);
            inputs[1] = new Matrix(nn.inputN, 1);
            inputs[2] = new Matrix(nn.inputN, 1);
            inputs[3] = new Matrix(nn.inputN, 1);

            targets[0] = new Matrix(nn.outputN, 1);
            targets[1] = new Matrix(nn.outputN, 1);
            targets[2] = new Matrix(nn.outputN, 1);
            targets[3] = new Matrix(nn.outputN, 1);

            inputs[0].matrix[0][0] = 0;
            inputs[0].matrix[1][0] = 0;
            inputs[1].matrix[0][0] = 0;
            inputs[1].matrix[1][0] = 1;
            inputs[2].matrix[1][0] = 1;
            inputs[2].matrix[0][0] = 0;
            inputs[3].matrix[0][0] = 1;
            inputs[3].matrix[1][0] = 1;

            targets[0].matrix[0][0] = 0;
            targets[1].matrix[0][0] = 1;
            targets[2].matrix[0][0] = 1;
            targets[3].matrix[0][0] = 0;

            for (int i = 0; i < epoc; i++)
            {
                for (int j = 0; j < inputs.length; j++)
                {
                    nn.input = inputs[j];
                    nn.target = targets[j];
                    nn.feedForward();
                    nn.backProp();
                }
            }


        }

    }
}